src/providers/tpm_provider/utils.rs

```diff
@@ -84,3 +91,134 @@ pub struct PasswordContext {
+            curve: convert_curve_to_tpm(curve_family, attributes.key_bits)?,
+        }),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+fn convert_asym_scheme_to_tpm(algorithm: Algorithm) -> Result<AsymSchemeUnion> {
+    match algorithm {
+        Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign { hash_alg }) => {
+            Ok(AsymSchemeUnion::RSASSA(convert_hash_to_tpm(hash_alg)?))
+        }
+        Algorithm::AsymmetricSignature(AsymmetricSignature::Ecdsa { hash_alg }) => {
+            Ok(AsymSchemeUnion::ECDSA(convert_hash_to_tpm(hash_alg)?))
+        }
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+#[allow(deprecated)]
+fn convert_hash_to_tpm(hash: Hash) -> Result<HashingAlgorithm> {
+    match hash {
+        Hash::Sha1 => Ok(HashingAlgorithm::Sha1),
+        Hash::Sha256 => Ok(HashingAlgorithm::Sha256),
+        Hash::Sha384 => Ok(HashingAlgorithm::Sha384),
+        Hash::Sha512 => Ok(HashingAlgorithm::Sha512),
+        Hash::Sha3_256 => Ok(HashingAlgorithm::Sha3_256),
+        Hash::Sha3_384 => Ok(HashingAlgorithm::Sha3_384),
+        Hash::Sha3_512 => Ok(HashingAlgorithm::Sha3_512),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+fn convert_curve_to_tpm(curve_family: EccFamily, key_size: u32) -> Result<EllipticCurve> {
+    match (curve_family, key_size) {
+        (EccFamily::SecpR1, 192) => Ok(EllipticCurve::NistP192),
+        (EccFamily::SecpR1, 224) => Ok(EllipticCurve::NistP224),
+        (EccFamily::SecpR1, 256) => Ok(EllipticCurve::NistP256),
+        (EccFamily::SecpR1, 384) => Ok(EllipticCurve::NistP384),
+        (EccFamily::SecpR1, 512) => Ok(EllipticCurve::NistP521),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+pub fn pub_key_to_bytes(pub_key: PublicKey, p_byte_size: usize) -> Result<Vec<u8>> {
+    match pub_key {
+        PublicKey::Rsa(key) => picky_asn1_der::to_vec(&RsaPublicKey {
+            modulus: IntegerAsn1::from_unsigned_bytes_be(key),
+            public_exponent: IntegerAsn1::from_signed_bytes_be(PUBLIC_EXPONENT.to_vec()),
+        })
+        .or(Err(ResponseStatus::PsaErrorGenericError)),
+        PublicKey::Ecc { x, y } => {
+            if x.len() != p_byte_size || y.len() != p_byte_size {
+                error!(
+                    "Received ECC public key with invalid size: x - {} bytes; y - {} bytes",
+                    x.len(),
+                    y.len()
+                );
+                return Err(ResponseStatus::PsaErrorGenericError);
+            }
+            Ok(elliptic_curve_point_to_octet_string(x, y))
+        }
+    }
+}
+
+fn elliptic_curve_point_to_octet_string(mut x: Vec<u8>, mut y: Vec<u8>) -> Vec<u8> {
+    let mut octet_string = vec![0x04];
+    octet_string.append(&mut x);
+    octet_string.append(&mut y);
+    octet_string
+}
+
+pub fn signature_data_to_bytes(data: SignatureData, p_byte_size: usize) -> Result<Vec<u8>> {
```

Comment:
The `p_byte_size` parameter is only used for the ECDSA signature arm. Would it be better if a `KeyAttributes` parameter type is passed instead (to mirror the `bytes_to_signature_data`) function and `p_byte_size` computed from `key_bits`?
That function could also check if the key type is matching the signature type and return `InvalidArgument`, same way as the other function.

Reply:
Also, this function is public while the other one is not, any reason for that?

Reply:
> Also, this function is public while the other one is not, any reason for that?

Only made the ones that are needed public - `utils` is not public past the provider anyway, so it's all limited to within the provider module.

Reply:
> That function could also check if the key type is matching the signature type and return InvalidArgument, same way as the other function.

Not sure if that's sensible, but could do. The `SignatureData` is coming from the TPM and there is already a conversion step there, from the raw FFI data to this.

Reply:
Yes I don't know if it would be useful then, apart from making sure that `data` is of the form we expect. I am just thinking that the caller would wonder what put for the `p_byte_size` parameter if doing RSA

Reply:
> I am just thinking that the caller would wonder what put for the p_byte_size parameter if doing RSA

Yeah, that makes sense!!